File: FlowCrypt/src/androidTest/java/com/flowcrypt/email/util/gson/MsgBlockAdapter.java
/*
 * © 2016-2019 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.util.gson;

import com.flowcrypt.email.api.retrofit.response.model.node.BaseMsgBlock;
import com.flowcrypt.email.api.retrofit.response.model.node.DecryptErrorMsgBlock;
import com.flowcrypt.email.api.retrofit.response.model.node.MsgBlock;
import com.flowcrypt.email.api.retrofit.response.model.node.PublicKeyMsgBlock;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

/**
 * @author Denis Bondarenko
 * Date: 3/19/19
 * Time: 2:58 PM
 * E-mail: DenBond7@gmail.com
 */
public class MsgBlockAdapter implements JsonDeserializer<MsgBlock> {
  @Override
  public MsgBlock deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    JsonObject jsonObject = json.getAsJsonObject();

    MsgBlock.Type type = context.deserialize(jsonObject.get("type"), MsgBlock.Type.class);

    if (type == null) {
      return null;
    }

    switch (type) {
      case PUBLIC_KEY:
        return context.deserialize(json, PublicKeyMsgBlock.class);

      case DECRYPT_ERROR:
        return context.deserialize(json, DecryptErrorMsgBlock.class);

      default:
        return context.deserialize(json, BaseMsgBlock.class);
    }
  }
}

File: FlowCrypt/src/androidTest/java/com/flowcrypt/email/util/gson/MsgBlockAdapter.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.util.gson

import com.flowcrypt.email.api.retrofit.response.model.DecryptErrorMsgBlock
import com.flowcrypt.email.api.retrofit.response.model.DecryptedAttMsgBlock
import com.flowcrypt.email.api.retrofit.response.model.GenericMsgBlock
import com.flowcrypt.email.api.retrofit.response.model.InlineAttMsgBlock
import com.flowcrypt.email.api.retrofit.response.model.MsgBlock
import com.flowcrypt.email.api.retrofit.response.model.PublicKeyMsgBlock
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import java.lang.reflect.Type

/**
 * @author Denys Bondarenko
 */
class MsgBlockAdapter : JsonDeserializer<MsgBlock> {
  override fun deserialize(
    json: JsonElement,
    typeOfT: Type,
    context: JsonDeserializationContext
  ): MsgBlock? {
    val jsonObject = json.asJsonObject

    val type = context.deserialize<MsgBlock.Type>(jsonObject.get("type"), MsgBlock.Type::class.java)
      ?: return null

    return when (type) {
      MsgBlock.Type.PUBLIC_KEY -> context.deserialize<MsgBlock>(
        json,
        PublicKeyMsgBlock::class.java
      )

      MsgBlock.Type.DECRYPTED_ATT -> context.deserialize<MsgBlock>(
        json,
        DecryptedAttMsgBlock::class.java
      )

      MsgBlock.Type.DECRYPT_ERROR -> context.deserialize<MsgBlock>(
        json,
        DecryptErrorMsgBlock::class.java
      )

      MsgBlock.Type.INLINE_ATT -> context.deserialize<MsgBlock>(
        json,
        InlineAttMsgBlock::class.java
      )

      else -> context.deserialize<MsgBlock>(json, GenericMsgBlock::class.java)
    }
  }
}
